### ISSUE TITLE
[Bug]: Flatpak doesn't capture TZDIR (assumes `/usr/share/zoneinfo` location), wrong time zone

### ISSUE THREAD

**Issue by podiki** [flatpak/flatpak#4715]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/master/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.12.3

### What Linux distribution are you using?

Other (specify below)

### Linux distribution version

GNU Guix

### What architecture are you using?

x86_64

### How to reproduce

1. Open an application, like us.zoom.Zoom and see incorrect time (it shows the raw UTC time rather than time zone corrected), or
2. Run a shell with `flatpak --user run --command=bash us.zoom.Zoom`
3. Run `date`

### Expected Behavior

Applications and `date` should have the correct time zone, e.g. `Jan 31 01:54:40 PM EST 2022` from this example.

### Actual Behavior

See the UTC (the system raw clock) time but with the wrong zone info, e.g. output is `Mon Jan 31 06:54:40 PM America 2022`. Note the "America" in the output indicating the zone is not properly read/used. Applications like Zoom show the wrong time in the main window. Other applications may do this differently, but the `date` command shows the base problem.

### Additional Information

This is on GNU Guix which does not have the usual `/usr/share/zoneinfo` but is kept in the store (similar to Nix). So on my machine I can see `TZDIR=/gnu/store/jj5aj3zxf65b7zgismwgy2wad7kbkcqz-tzdata-2021e/share/zoneinfo` where that long string is the hash of the package. I also have `TZ` set correctly and both `/etc/localtime` and `/etc/timezone` set properly.

The problem is that Flatpak keeps the `TZDIR` variable but does not capture that directory in the environment so it cannot read the zone info. I believe commit dc4e198766d91aa8871c276b9ed8d7f3ba003822 has an assumption that `/usr/share/zoneinfo` is where to look, rather than relying on `TZDIR` and capturing that directory.

This can be worked around on the user end by running flatpak with `TZDIR= flatpak` to unset it, or use `TZDIR=/usr/share/zoneinfo` so it is read within the container, or with `--filesystem=$TZDIR` to properly expose the `TZDIR`. Each of these gives me the proper output of `date` and in e.g. us.zoom.Zoom.

But I believe this is something that Flatpak should account for, not assuming where zoneinfo is, or at least making sure it can properly read the zoneinfo. I think the simplest would be to `getenv` for `TZDIR`, defaulting to `/usr/share/zoneinfo` if this is not set or the directory cannot be read/shared to the container.

**Comment by nanonyme** [comment]
Looks like based on flatpak source code that in fact does in some cases try to use TZDIR https://github.com/flatpak/flatpak/blob/c392e5adf63ea2817a6f05081365fdd7acdfc04e/common/flatpak-utils-base.c#L48

**Comment by podiki** [comment]
> Looks like based on flatpak source code that in fact does in some cases try to use TZDIR https://github.com/flatpak/flatpak/blob/c392e5adf63ea2817a6f05081365fdd7acdfc04e/common/flatpak-utils-base.c#L48

(I got this reply but do not see it on here...?)

Yes, some places use `TZDIR`, but critically not when finding what directory to expose, from dc4e198

**Comment by nanonyme** [comment]
Right, when someone deletes post, you will still retain email notification. It is sounds like a bug you can quoted reply. I did understand your problem later.

**Comment by claes** [comment]
To me it looks like https://github.com/flatpak/flatpak/pull/5850/ attempts to address this, but that it does not work. The PR was backported to NixOS / nixpkgs in https://github.com/NixOS/nixpkgs/pull/343590 but it does not give the desired effect. Would be good to understand how https://github.com/flatpak/flatpak/pull/5850/ is intended to work

**Comment by chrisawi** [comment]
I think this was ultimately fixed by https://github.com/flatpak/flatpak/pull/5942, at least for NixOS. Is there any reason to think Guix would work differently?

**Comment by smcv** [comment]
> I think this was ultimately fixed by #5942, at least for NixOS. Is there any reason to think Guix would work differently?

That seems plausible to me: NixOS and Guix are "the same shape" from Flatpak's point of view (the precise paths are different, but the general idea is the same).

@podiki, please could you re-test this with Flatpak 1.6.x, and leave a comment here or open a new issue if it is not fixed? (Or, open a merge request with a proposed solution.)

**Comment by podiki** [comment]
> @podiki, please could you re-test this with Flatpak 1.6.x, and leave a comment here or open a new issue if it is not fixed? (Or, open a merge request with a proposed solution.)

Thanks for the followup all. I just tested as I did in the original report and the time was correct (Flatpak 1.14.10 currently). Apologies for not responding earlier, luckily I can now use Guix packages just about exclusively.
